### PR TITLE
Attention

### PR DESCRIPTION
**Pull request by omarnmahmood**
Implementation of decomposable attention model and softmax criterion for training on [real, fake, fake] interactions

### CODE REVIEW

**Review by mortonjt — APPROVED** [review]
I think this is ok, although, I'm a little confused on what softmax means here - its not clear to me where `use_softmax` depends on a softmax.

**Comment by ehthiede** [comment]
For simplicity, should we move the routines on the softmax into a separate file, possibly one that imports form train_nce.py? ( We might need to better encapsulate train_nce for this...)  I feel like it's a sufficient deviation that it should receive it's own training script (both the dataset and the output are different).  It's more like BPR, and I think it makes more sense to have those in the same script.

**Comment by omarnmahmood** [comment]
> I think this is ok, although, I'm a little confused on what software means here - its not clear to me where `use_softmax` depends on a softmax.

I don't explicitly make a call to softmax because the cross-entropy loss function does that by default: https://pytorch.org/docs/stable/nn.html#crossentropyloss

**Comment by omarnmahmood** [comment]
> For simplicity, should we move the routines on the softmax into a separate file, possibly one that imports form train_nce.py? ( We might need to better encapsulate train_nce for this...) I feel like it's a sufficient deviation that it should receive it's own training script (both the dataset and the output are different). It's more like BPR, and I think it makes more sense to have those in the same script.

Sure that makes sense.  I can separate it into a different file

File: examples/train_softmax.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        pickle.dump(args, f, pickle.HIGHEST_PROTOCOL)
+
+    lm, path = pretrained_language_models[args.lmarch]
+
+    train_dataset = ComparisonDrugProteinDataset(args.train_dataset)
+    valid_dataset = ComparisonDrugProteinDataset(args.valid_dataset)
+
+    cfxn = lambda x: collate_fn_triplet(x, prots_are_sequences=True)
+    train_dataloader = DataLoader(train_dataset, args.batch_size, shuffle=True, collate_fn=cfxn)
+    valid_dataloader = DataLoader(valid_dataset, args.batch_size, shuffle=True, collate_fn=cfxn)
```
Comment (mortonjt) [inline examples/train_softmax.py]: I bet that we will need `drop_last=True` here, since the last batch in the epoch usually messes things up and changes the dimensions. https://pytorch.org/docs/stable/data.html

File: examples/train_softmax.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
+import torch
+import argparse
+import os
+import pickle
+import datetime
+from torch import nn
+from binding_prediction.models import BindingModel
+from binding_prediction.dataset import ComparisonDrugProteinDataset, collate_fn_triplet
+from binding_prediction import pretrained_language_models
+from examples.train_nce import get_targets, run_model_on_batch, initialize_logging
```
Comment (mortonjt) [inline examples/train_softmax.py]: I'm getting an error here when trying to run this code.  This seems to be resolved when I replace this with
```suggestion
from train_nce import get_targets, run_model_on_batch, initialize_logging
```
